File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConsent/JWT.swift
import Foundation

#if canImport(Security)
import Security
#endif

/// Internal JWT codec. Hand-rolled HS256 / RS256 using CryptoKit (HMAC)
/// and `Security.framework` (RSA on Apple platforms) or swift-crypto
/// (RSA on Linux). No third-party JWT library required.
public enum JWT {

    /// Parsed (but not verified) JWT.
    public struct Parsed {
        public let headerB64: String
        public let claimsB64: String
        public let signatureB64: String
        public let header: [String: Any]
        public let claims: [String: Any]
        public let signature: Data
    }

    /// Sign claims into a compact JWT.
    ///
    /// - Parameter hsKey: raw HMAC bytes (HS256). Required for HS256.
    /// - Parameter rsPrivateKey: PKCS#8 / PEM private key data (RS256).
    public static func sign(
        algorithm: Algorithm,
        claims: [String: Any],
        hsKey: Data? = nil,
        rsPrivateKey: Data? = nil
    ) throws -> String {
        let header: [String: Any] = ["alg": algorithm.rawValue, "typ": "JWT"]
        let headerJSON = try canonicalJSONData(header)
        let claimsJSON = try canonicalJSONData(claims)
        let headerB64 = Base64URL.encode(headerJSON)
        let claimsB64 = Base64URL.encode(claimsJSON)
        let signingInput = headerB64 + "." + claimsB64
        guard let signingBytes = signingInput.data(using: .utf8) else {
            throw ConsentError.invalidToken(reason: "could not encode signing input")
        }

        let signature: Data
        switch algorithm {
        case .hs256:
            guard let key = hsKey, !key.isEmpty else {
                throw ConsentError.invalidToken(reason: "HS256 requires hsKey")
            }
            signature = hmacSha256(message: signingBytes, key: key)
        case .rs256:
            guard let priv = rsPrivateKey else {
                throw ConsentError.invalidToken(reason: "RS256 requires rsPrivateKey")
            }
            signature = try rsaSignSha256(message: signingBytes, privateKey: priv)
        }
        return signingInput + "." + Base64URL.encode(signature)
    }

    /// Parse a compact JWT into its three deserialized parts.
    /// Does NOT verify the signature.
    public static func parse(_ token: String) throws -> Parsed {
        let parts = token.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3 else {
            throw ConsentError.invalidToken(reason: "malformed JWT: expected 3 segments")
        }
        let headerB64 = String(parts[0])
        let claimsB64 = String(parts[1])
        let sigB64 = String(parts[2])
        let headerData = try Base64URL.decode(headerB64)
        let claimsData = try Base64URL.decode(claimsB64)
        let sigData = try Base64URL.decode(sigB64)

        let headerObj = try JSONSerialization.jsonObject(with: headerData)
        guard let header = headerObj as? [String: Any] else {
            throw ConsentError.invalidToken(reason: "JWT header must be an object")
        }
        let claimsObj = try JSONSerialization.jsonObject(with: claimsData)
        guard let claims = claimsObj as? [String: Any] else {
            throw ConsentError.invalidToken(reason: "JWT claims must be an object")
        }
        return Parsed(
            headerB64: headerB64,
            claimsB64: claimsB64,
            signatureB64: sigB64,
            header: header,
            claims: claims,
            signature: sigData)
    }

    /// Verify the signature on a parsed JWT.
    public static func verifySignature(
        _ parsed: Parsed,
        algorithm: Algorithm,
        hsKey: Data? = nil,
        rsPublicKey: Data? = nil
    ) throws -> Bool {
        let signingInput = parsed.headerB64 + "." + parsed.claimsB64
        guard let signingBytes = signingInput.data(using: .utf8) else {
            return false
        }
        switch algorithm {
        case .hs256:
            guard let key = hsKey else { return false }
            let expected = hmacSha256(message: signingBytes, key: key)
            return constantTimeEquals(expected, parsed.signature)
        case .rs256:
            guard let pub = rsPublicKey else { return false }
            return try rsaVerifySha256(
                message: signingBytes, signature: parsed.signature, publicKey: pub)
        }
    }

    // MARK: - Internals

    /// Encode a Foundation JSON value with stable key order. The IAIso JWT
    /// claim format requires a specific field order (iss, sub, iat, exp,
    /// jti, scopes, ...) — the caller controls order via an ordered map
    /// passed via the `orderedKeys` parameter, OR we sort alphabetically
    /// for determinism in the absence of explicit ordering.
    static func canonicalJSONData(_ value: [String: Any]) throws -> Data {
        // For interoperability with other reference SDKs, key order is NOT
        // sorted alphabetically — `Issuer.issue()` builds the claims dict
        // in the spec order and we encode preserving that order.
        // However, JSONSerialization on Apple platforms doesn't preserve
        // key order from a Swift dictionary. We hand-build the JSON using
        // a deterministic but order-preserving approach: the caller passes
        // a `[String: Any]` whose iteration order is undefined; we encode
        // using JSONSerialization (which IS stable for primitive-only maps
        // in practice on Apple platforms but not guaranteed). For maximum
        // safety, callers that care about wire-format order should call
        // `canonicalJSONDataOrdered` with an explicit ordered list.
        return try JSONSerialization.data(
            withJSONObject: value,
            options: [.sortedKeys, .withoutEscapingSlashes])
    }

    /// Hand-build canonical JSON from an explicit ordered list of (key, value).
    /// Used by `Issuer` to emit claims in spec order.
    static func canonicalJSONDataOrdered(_ pairs: [(String, Any)]) throws -> Data {
        var s = "{"
        for (i, (k, v)) in pairs.enumerated() {
            if i > 0 { s += "," }
            s += jsonEncodeString(k) + ":"
            s += try jsonEncode(v)
        }
        s += "}"
        guard let data = s.data(using: .utf8) else {
            throw ConsentError.invalidToken(reason: "encode failed")
        }
        return data
    }

    static func jsonEncodeString(_ s: String) -> String {
        // Use JSONSerialization for proper escaping by encoding via a 1-element array.
        if let data = try? JSONSerialization.data(
            withJSONObject: [s], options: [.withoutEscapingSlashes]),
           let str = String(data: data, encoding: .utf8) {
            // Strip surrounding [ ]
            var t = str
            if t.hasPrefix("["), t.hasSuffix("]") {
                t = String(t.dropFirst().dropLast())
            }
            return t
        }
        return "\"\(s)\""
    }

    static func jsonEncode(_ value: Any) throws -> String {
        if value is NSNull { return "null" }
        // NSNumber check first to disambiguate bool vs numeric.
        if let n = value as? NSNumber {
            #if canImport(ObjectiveC)
            if String(cString: n.objCType) == "c" {
                return n.boolValue ? "true" : "false"
            }
            #endif
            // Numeric.
            let d = n.doubleValue
            if d.truncatingRemainder(dividingBy: 1) == 0, abs(d) < 1e16 {
                return String(n.int64Value)
            }
            return String(d)
        }
        if let b = value as? Bool, type(of: value) == Bool.self {
            return b ? "true" : "false"
        }
        if let i = value as? Int { return String(i) }
        if let i = value as? Int64 { return String(i) }
        if let d = value as? Double {
            if d.truncatingRemainder(dividingBy: 1) == 0, abs(d) < 1e16 {
                return String(Int64(d))
            }
            return String(d)
        }
        if let s = value as? String { return jsonEncodeString(s) }
        if let arr = value as? [Any] {
            let parts = try arr.map { try jsonEncode($0) }
            return "[" + parts.joined(separator: ",") + "]"
        }
        if let m = value as? [String: Any] {
            let data = try JSONSerialization.data(
                withJSONObject: m,
                options: [.sortedKeys, .withoutEscapingSlashes])
            return String(data: data, encoding: .utf8) ?? "{}"
        }
        // Fallback: use JSONSerialization
        let data = try JSONSerialization.data(
            withJSONObject: [value],
            options: [.withoutEscapingSlashes])
        var s = String(data: data, encoding: .utf8) ?? ""
        if s.hasPrefix("["), s.hasSuffix("]") {
            s = String(s.dropFirst().dropLast())
        }
        return s
    }

    // MARK: - Crypto primitives

    static func hmacSha256(message: Data, key: Data) -> Data {
        let symKey = SymmetricKey(data: key)
        let mac = HMAC<SHA256>.authenticationCode(for: message, using: symKey)
        return Data(mac)
    }

    static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for i in 0..<a.count {
            diff |= a[i] ^ b[i]
        }
        return diff == 0
    }

    static func rsaSignSha256(message: Data, privateKey: Data) throws -> Data {
        #if canImport(Security)
        let key = try parsePrivateKey(privateKey)
        var error: Unmanaged<CFError>?
        guard let sig = SecKeyCreateSignature(
            key,
            .rsaSignatureMessagePKCS1v15SHA256,
            message as CFData,
            &error) as Data? else {
            let msg = error?.takeRetainedValue().localizedDescription ?? "RSA sign failed"
            throw ConsentError.invalidToken(reason: msg)
        }
        return sig
        #else
        throw ConsentError.invalidToken(
            reason: "RSA signing not implemented on this platform")
        #endif
    }

    public static func rsaVerifySha256(
        message: Data, signature: Data, publicKey: Data
    ) throws -> Bool {
        #if canImport(Security)
        let key = try parsePublicKey(publicKey)
        var error: Unmanaged<CFError>?
        let ok = SecKeyVerifySignature(
            key,
            .rsaSignatureMessagePKCS1v15SHA256,
            message as CFData,
            signature as CFData,
            &error)
        return ok
        #else
        return false
        #endif
    }

    #if canImport(Security)
    /// Parse a PEM- or DER-encoded RSA private key into a SecKey.
    static func parsePrivateKey(_ pemOrDer: Data) throws -> SecKey {
        let der = try stripPemHeader(pemOrDer)
        let attrs: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPrivate,
        ]
        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(der as CFData, attrs as CFDictionary, &error) else {
            let msg = error?.takeRetainedValue().localizedDescription ?? "bad RSA private key"
            throw ConsentError.invalidToken(reason: msg)
        }
        return key
    }

    /// Parse a PEM- or DER-encoded RSA public key into a SecKey.
    static func parsePublicKey(_ pemOrDer: Data) throws -> SecKey {
        let der = try stripPemHeader(pemOrDer)
        let attrs: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
        ]
        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(der as CFData, attrs as CFDictionary, &error) else {
            let msg = error?.takeRetainedValue().localizedDescription ?? "bad RSA public key"
            throw ConsentError.invalidToken(reason: msg)
        }
        return key
    }

    /// Strip PEM headers and base64-decode the body. Passes raw DER through.
    static func stripPemHeader(_ data: Data) throws -> Data {
        guard let s = String(data: data, encoding: .utf8) else { return data }
        if !s.contains("-----BEGIN") { return data }  // already DER
        let lines = s.components(separatedBy: .newlines)
            .filter { !$0.hasPrefix("-----") && !$0.isEmpty }
        let b64 = lines.joined()
        guard let der = Data(base64Encoded: b64) else {
            throw ConsentError.invalidToken(reason: "could not decode PEM body")
        }
        return der
    }
    #endif
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/consent/jwt.rb
# frozen_string_literal: true

require "openssl"
require "base64"
require "json"
require_relative "errors"
require_relative "algorithm"

module IAIso
  module Consent
    # Internal JWT codec. Hand-rolled HS256/RS256 using OpenSSL stdlib.
    # No third-party JWT library required.
    module JWT
      module_function

      # Sign claims into a compact JWT.
      #
      # @param algorithm [String] Algorithm::HS256 or Algorithm::RS256
      # @param claims [Hash] insertion order is preserved (Ruby 1.9+)
      # @param hs_key [String, nil] raw HMAC bytes (HS256)
      # @param rs_private_key [String, nil] PEM-encoded RSA private key (RS256)
      def sign(algorithm:, claims:, hs_key: nil, rs_private_key: nil)
        header = { "alg" => algorithm, "typ" => "JWT" }
        header_json = canonical_json(header)
        # Ruby preserves Hash insertion order, so JSON.generate preserves
        # the spec-mandated claim order.
        claims_json = canonical_json(claims)
        header_b64 = b64url_encode(header_json)
        claims_b64 = b64url_encode(claims_json)
        signing_input = "#{header_b64}.#{claims_b64}"
        signature =
          case algorithm
          when Algorithm::HS256
            raise InvalidTokenError, "HS256 requires hs_key" if hs_key.nil? || hs_key.empty?
            OpenSSL::HMAC.digest("sha256", hs_key, signing_input)
          when Algorithm::RS256
            raise InvalidTokenError, "RS256 requires rs_private_key" if rs_private_key.nil?
            key = rs_private_key.is_a?(OpenSSL::PKey::RSA) ? rs_private_key : OpenSSL::PKey::RSA.new(rs_private_key)
            key.sign(OpenSSL::Digest.new("SHA256"), signing_input)
          else
            raise InvalidTokenError, "unsupported algorithm: #{algorithm}"
          end
        "#{signing_input}.#{b64url_encode(signature)}"
      end

      # Parse a compact JWT into its three parts. Does NOT verify the signature.
      def parse(token)
        parts = token.split(".", -1)
        unless parts.size == 3
          raise InvalidTokenError, "malformed JWT: expected 3 segments"
        end
        header_b64, claims_b64, sig_b64 = parts
        header = ::JSON.parse(b64url_decode(header_b64))
        claims = ::JSON.parse(b64url_decode(claims_b64))
        signature = b64url_decode(sig_b64)
        unless header.is_a?(Hash) && claims.is_a?(Hash)
          raise InvalidTokenError, "JWT header/claims must be objects"
        end
        {
          header_b64: header_b64,
          claims_b64: claims_b64,
          signature_b64: sig_b64,
          header: header,
          claims: claims,
          signature: signature,
        }
      rescue ::JSON::ParserError => e
        raise InvalidTokenError, "malformed JWT JSON: #{e.message}"
      end

      # Verify the signature on a parsed JWT.
      def verify_signature(parsed, algorithm:, hs_key: nil, rs_public_key: nil)
        signing_input = "#{parsed[:header_b64]}.#{parsed[:claims_b64]}"
        case algorithm
        when Algorithm::HS256
          return false if hs_key.nil?
          expected = OpenSSL::HMAC.digest("sha256", hs_key, signing_input)
          # constant-time compare via OpenSSL.fixed_length_secure_compare
          return false unless expected.bytesize == parsed[:signature].bytesize
          OpenSSL.fixed_length_secure_compare(expected, parsed[:signature])
        when Algorithm::RS256
          return false if rs_public_key.nil?
          key = rs_public_key.is_a?(OpenSSL::PKey::RSA) ? rs_public_key : OpenSSL::PKey::RSA.new(rs_public_key)
          key.verify(OpenSSL::Digest.new("SHA256"), parsed[:signature], signing_input)
        else
          false
        end
      end

      # Encode a Hash to JSON, preserving insertion order (Ruby 1.9+).
      # We don't escape forward slashes — matches the other reference SDKs.
      def canonical_json(value)
        ::JSON.generate(value, ascii_only: false)
      end

      def b64url_encode(bytes)
        Base64.urlsafe_encode64(bytes, padding: false)
      end

      def b64url_decode(str)
        # Ruby's urlsafe_decode64 is strict about padding; pad before decoding.
        padded = str.dup
        rem = padded.length % 4
        padded << ("=" * (4 - rem)) if rem.positive?
        Base64.urlsafe_decode64(padded)
      rescue ArgumentError => e
        raise InvalidTokenError, "malformed base64url segment: #{e.message}"
      end
    end
  end
end
